chain/chain/src/chain.rs

```diff
@@ -1464,53 +1457,91 @@ impl Chain {
-                    {
-                        let mut count = 0;
-                        let mut stake = 0;
-                        for (info, is_slashed) in producers.iter() {
-                            if !*is_slashed {
-                                stake += info.stake();
-                                count += 1;
+        for (block, block_preprocess_info, apply_results) in applied_blocks {
+            let prev_head = self.store.head()?;
+            let mut chain_update = self.chain_update();
+            let res =
+                chain_update.postprocess_block(me, &block, block_preprocess_info, apply_results);
+            match res {
+                Ok(head) => {
+                    chain_update
+                        .chain_store_update
+                        .save_block_height_processed(block.header().height());
+                    chain_update.commit()?;
+
+                    self.pending_states_to_patch = None;
+
+                    if let Some(tip) = &head {
+                        if let Ok(producers) = self
+                            .runtime_adapter
+                            .get_epoch_block_producers_ordered(&tip.epoch_id, &tip.last_block_hash)
+                        {
+                            let mut count = 0;
+                            let mut stake = 0;
+                            for (info, is_slashed) in producers.iter() {
+                                if !*is_slashed {
+                                    stake += info.stake();
+                                    count += 1;
+                                }
                             }
+                            stake /= NEAR_BASE;
+                            metrics::VALIDATOR_AMOUNT_STAKED
+                                .set(i64::try_from(stake).unwrap_or(i64::MAX));
+                            metrics::VALIDATOR_ACTIVE_TOTAL.set(count);
                         }
-                        stake /= NEAR_BASE;
-                        metrics::VALIDATOR_AMOUNT_STAKED
-                            .set(i64::try_from(stake).unwrap_or(i64::MAX));
-                        metrics::VALIDATOR_ACTIVE_TOTAL.set(count);
-                    }
-                };
+                    };
 
-                let status = self.determine_status(head.clone(), prev_head);
+                    let status = self.determine_status(head.clone(), prev_head);
 
-                // Notify other parts of the system of the update.
-                block_accepted(AcceptedBlock { hash: *block.hash(), status, provenance });
+                    // Notify other parts of the system of the update.
+                    block_accepted(AcceptedBlock {
+                        hash: *block.hash(),
+                        status,
+                        provenance: provenance.clone(),
+                    });
 
-                Ok(head)
+                    if head.is_some() {
+                        maybe_new_head = head;
```

Comment:
what should we do if there were multiple blocks ? 
with the current code, we'd set the new head to be the head of the latest one - but is it the right thing to do?

Reply:
The current code actually does something different, it sets the new head to be the head of the latest one that is not empty, which I think is right. 

Any way, I got rid of that code for now. I'll take care of the logic in the next PR. The tricky part is error handling of multiple blocks.